### PR TITLE
[FIX] udes_priority: Fixes for priority propagation

### PR DESCRIPTION
**Pull request by joe-gb**
User-story: 11723

### CODE REVIEW

File: addons/udes_priorities/models/stock_move.py [hunk 1]
```diff
@@ -39,4 +43,33 @@ def _priority_domain(self):
     def get_priorities_for_selection(self):
         Priorities = self.env["udes_priorities.priority"]
         priorities = Priorities.search(self._priority_domain())
+
+        # hard coded default value means there is always a priority to set
+        normal = self.env.ref("udes_priorities.normal")
+        priorities |= normal
```
Comment (ws-mps) [inline addons/udes_priorities/models/stock_move.py:49]: Do we want to do this only if `priorities` is empty?

File: addons/udes_priorities/models/stock_move.py [hunk 2]
```diff
@@ -39,4 +43,33 @@ def _priority_domain(self):
...
+
+        for move in self.filtered(lambda m: m.priority not in priorities.mapped("reference")):
+            # Theres some race conditions around where data is aviable to search on this means
```
Comment (ws-mps) [inline addons/udes_priorities/models/stock_move.py:52]: Typo: aviable

File: addons/udes_priorities/models/stock_picking.py [hunk 3]
```diff
@@ -37,31 +40,41 @@ def _priority_and_priority_group_domain(self):
...
             active_pick = self.browse(active_id).exists()
             if active_pick:
                 pick = active_pick
 
-        return Priorities.search(pick._priority_domain()).get_selection_values()
+        priorities = Priorities.search(pick._priority_domain())
+
+        # hard coded default value means there is always a priority to set
+        normal = self.env.ref("udes_priorities.normal")
+        priorities |= normal
```
Comment (ws-mps) [inline addons/udes_priorities/models/stock_picking.py:67]: Same as above.

File: addons/udes_priorities/models/stock_picking.py [hunk 4]
```diff
@@ -81,13 +94,23 @@ def get_priorities(self):
...
-        for group in PriorityGroup.search(self._priority_group_domain()):
-            groups.append(
-                {
-                    "name": group.name,
-                    "priorities": [{"id": p.reference, "name": p.name} for p in group.priority_ids],
-                }
-            )
+        group_domain_kwargs = {}
+        if picking_type_id is None:
+            group_domain_kwargs["picking_type_ids"] = [picking_type_id]
```
Comment (ws-mps) [inline addons/udes_priorities/models/stock_picking.py:99]: `picking_type_id` is None here, so why don't just []?

Comment (unirho) [reply addons/udes_priorities/models/stock_picking.py:99]: should probably be not none but mick wasn't sure why he had put this

File: addons/udes_priorities/models/stock_picking.py (hunk 3 above)
Comment (unirho) [reply addons/udes_priorities/models/stock_picking.py:67]: as we can default to normal we always want to provide that back as a choice, we could run into some nasties where move has defaulted to normal (say we make one without pick type) and the picking we try to give it to doesn't have normal as an acceptable value. This is mostly relevant when you try to set values to nothing, there has to be a supreme default, along the lines of why vanilla odoo always puts 1

File: addons/udes_priorities/tests/test_stock_picking.py [hunk 5]
```diff
@@ -23,3 +23,31 @@ def test_correct_picking_type_or_no_picking_type_is_valid(self):
...
+        Move = self.env["stock.move"]
+        apple = Product.create(
+            {"name": "Apple", "barcode": "bar-apple", "default_code": "APL", "type": "product",}
+        )
+        self.not_urgent.picking_type_ids = self.picking_type_pick
+        self.assertEqual(self.not_urgent.picking_type_ids, self.picking_type_pick)
+        pick = Picking.create({"priority": self.not_urgent.reference})
+        self.assertEqual(pick.priority, self.not_urgent.reference)
+        move = Move.create(
+            {
```
Comment (unirho) [inline addons/udes_priorities/tests/test_stock_picking.py:41]: as you see here we provide nothing for the priority for the move and so it is defaulted to normal, which means we must include that in the selection domain defined by the method "get_priorities_for_selection"

File: addons/udes_priorities/models/stock_picking.py (hunk 4 above)
Comment (unirho) [reply addons/udes_priorities/models/stock_picking.py:99]: i will update this in the hht story so we can test it as that is what this function is intended for, as you say it doesn't block this

File: addons/udes_priorities/models/stock_move.py (hunk 1 above)
Comment (unirho) [reply addons/udes_priorities/models/stock_move.py:49]: resolving as per comments

**Review by unirho — APPROVED** [review]

File: addons/udes_priorities/models/stock_move.py [hunk 6]
```diff
@@ -39,4 +43,33 @@ def _priority_domain(self):
...
+            # Theres some race conditions around where data is available to search on this means
+            # that sometimes an invalid default can be set
+            move.priority = normal.reference
+
         return priorities.get_selection_values()
+
+    @api.constrains("priority")
+    @api.depends("priority")
+    @api.onchange("priority")
+    def _priority_cant_be_empty(self):
```
Comment (apatrick88) [inline addons/udes_priorities/models/stock_move.py:61]: Strictly speaking, do we need all of constrains, depends and onchange here? Would we get what we need with just constrains?

Comment (joe-gb) [reply addons/udes_priorities/models/stock_move.py:61]: constrains and onchange are needed, I think depends is the one we can remove... but not 100%

Comment (joe-gb) [reply addons/udes_priorities/models/stock_move.py:60]: removed :)

**Review by apatrick88 — APPROVED** [review]

**Pull request merged**